### PR TITLE
Google Analytics: Add Support for Custom Dimensions and Metrics

### PR DESCRIPTION
**Pull request by sperand-io**
- [x] time with @f2prateek to go over and get 20% review
- [x] get it so this isn't breaking any existing tests
- [x] write tests for CDs and CMs
- [x] get those passing 
- [ ] final review by @f2prateek

### CODE REVIEW

File: analytics-integrations/google-analytics/src/main/java/com/segment/analytics/internal/integrations/GoogleAnalyticsIntegration.java [hunk 1]
```diff
@@ -166,6 +202,20 @@ boolean handleProductEvent(String event, String category, Properties properties)
     return false;
   }
 
+  void setCustomDimensions(HitBuilders.HitBuilder hit, Properties properties) {
```
Comment (sperand-io) [inline analytics-integrations/google-analytics/src/main/java/com/segment/analytics/internal/integrations/GoogleAnalyticsIntegration.java]: looks like HitBuilders.HitBuilder is protected, so I'm not sure of how to abstract this out, may just have to have some repetitive logic in `.screen` and `.track`

File: analytics-integrations/google-analytics/src/main/java/com/segment/analytics/internal/integrations/GoogleAnalyticsIntegration.java [hunk 2]
```diff
@@ -73,6 +76,16 @@
     }
     // tracker.setSampleRate(settings.getDouble("siteSpeedSampleRate", 1));
     sendUserId = settings.getBoolean("sendUserId", false);
+
+    customDimensions = settings.getValueMap("dimensions");
```
Comment (f2prateek) [inline analytics-integrations/google-analytics/src/main/java/com/segment/analytics/internal/integrations/GoogleAnalyticsIntegration.java]: This should already be an empty map coming from the server? No need to have the `null` check below.

File: analytics-integrations/google-analytics/src/main/java/com/segment/analytics/internal/integrations/GoogleAnalyticsIntegration.java (hunk 1 above)
Comment (f2prateek) [inline analytics-integrations/google-analytics/src/main/java/com/segment/analytics/internal/integrations/GoogleAnalyticsIntegration.java]: We could create two new classes `EventHitBuilder` and `AppHitViewBuilder` that extend `EventBuilder` and `AppViewBuilder` respectively, and implement a new interface `CustomHitBuilder` that this method accepts. 

`void setCustomDimensions(CustomHitBuilder hit, Properties properties)`

```java
  interface CustomHitBuilder {
    CustomHitBuilder setCustomDimension(int index, String dimension);
  }

  static class EventHitBuilder extends HitBuilders.EventBuilder implements CustomHitBuilder {
    @Override public EventHitBuilder setCustomDimension(int index, String dimension) {
      super.setCustomDimension(index, dimension);
      return this;
    }
  }

  static class AppViewHitBuilder extends HitBuilders.EventBuilder implements CustomHitBuilder {
    @Override public AppViewHitBuilder setCustomDimension(int index, String dimension) {
      super.setCustomDimension(index, dimension);
      return this;
    }
  }
```

Comment (sperand-io) [inline analytics-integrations/google-analytics/src/main/java/com/segment/analytics/internal/integrations/GoogleAnalyticsIntegration.java]: nice! thanks @f2prateek :)

**Pull request closed without merging**